BEPs/BEP-396.md

```diff
@@ -0,0 +1,187 @@
+
+If a transaction reads and writes the same value at a memory location, the location can be considered read-only.
+
+### 5.2 TxDAG Structure
+
+The structure of `TxDAG` is defined as follows.
+
+```go
+type TxDep struct {
+	// It describes the Relation with below txs 
+	// 0: this tx depends on below txs 
+	// 1: this transaction does not depend on below txs, all other previous txs depend on
+	Relation  uint8
+	TxIndexes []int
+}
+
+type TxDAG struct {
+	// The TxDAG type
+	// 0: delay the distribution of GasFee, it will ignore all gas fee distribution when tx execute
+	// 1: timely distribution of transaction fees, it will keep partial serial execution when tx cannot delay the distribution
+	Type uint8
+	// Tx Dependency List, the list index is equal to TxIndex
+	TxDeps []TxDep
+	// It indicates the scheduling priority of the transactions
+	SchedulePriority []int
+}
+```
+
+`SchedulePriority` is used to sort transactions by scheduling priority, allowing the scheduler to prioritize the most time-consuming parallel paths, trying to improve the scheduling efficiency under limited CPU resources. This is optional and the default is to sort by TxIndex.
+
+There is also a potential `TxDAG` "compression algorithm" to further optimize the size. For the DAG describing the dependencies, some are redundant because it only needs to ensure that the final execution path is consistent.
+
+![](./assets/BEP-396/pic2.png)
+
+Specific rules: If Txj depends on Txi, j>i, then Txj can ignore the same dependencies as Txi, and the execution path of Txj after the update remains the same as before.
+
+This can keep the size of `TxDAG` as small as possible and avoid generating too large `TxDAG` and wasting network bandwidth.
+
+### 5.3 TxDAG Validation
+
+Since there is a preset order of transactions in the block, Txi can't depend on Txj (i<j) in DAG, and it is also impossible for circular dependencies to occur. 
+
+The verification rules are as follows:
+
+1. TxIndexes must be smaller than the current TxIndex;
+2. TxIndexes cannot be repeated;
+3. TxIndexes must be sorted in ascending order;
+4. TxIndex in `TxDAG` cannot exceed the maximum TxIndex of the block;
+5. Verify whether the dependency optimization rules are followed and delete redundant dependencies;
+
+### 5.4 Block Extension
+
+This BEP stores `TxDAG` data in the `BlockBody`, and its Hash in the Extra field of the `Header`. To ensure the correctness of the DAG, if the validator finds an incorrect DAG, it should reject the block.
+
+This is a diagram of the Extra field structure changes in the `Header`:
+
+![](assets/BEP-396/pic3.png)
+
+The specific `BlockBody` structure is defined as follows:
+
+```go
+type Body struct {
+	Transactions []*Transaction
+	Uncles       []*Header
+	Withdrawals  []*Withdrawal `rlp:"optional"`
+	TxDAG        *TxDAG        `rlp:"optional"`
```

Comment:
This will break the web3.js and other SDK and tools, what is the plan to deal with it?

Reply:
If the user uses the latest BSC SDK, everything is fine. If the user uses the Eth ecological tool or the old SDK, TxDAG may not be parsed because TxDAG will always remain at the end of the rlp encoding. Of course, it is recommended that users use the latest BSC SDK.

Reply:
This field will greatly impact the hash calculation of a block. You know a lot of off chain index heavyly depdens on the hash of a block, so for a Eth ecological tool or the old SDK, they will get a different blockhash.

Reply:
Oh, the user also can generate the correct hash by header, The DAG Hash has been stored in extra field of header, but it may cannot verify the TxDAG data if they use old SDK or eth's tool.